### ISSUE TITLE
Error: Unexpected top-level property "$schema"

### ISSUE THREAD

**Issue by octogonz** [eslint/eslint#12010]
**Tell us about your environment**

* **ESLint Version:** 6.0.1
* **Node Version:** 10.15.0
* **npm Version:** 6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?** `@typescript-eslint/parser`

**Please show your full configuration:**

```js
{
  "$schema": "http://json.schemastore.org/eslintrc",
  "parser": "@typescript-eslint/parser",
  "plugins": [
    "@typescript-eslint"
  ],
  "parserOptions": {
    "project": "./tsconfig.json"
  },
  "rules": {
    "@typescript-eslint/restrict-plus-operands": "error"
  }
}
```

**What did you expect to happen?**

The `$schema` field is the standard way to tell editors where to find schema for a JSON file.  Without this field, IntelliSense won't be available when people edit the **.eslint.json** config file.

The `$schema` field is documented in the JSON Schema spec:
https://json-schema.org/latest/json-schema-core.html#rfc.section.7

**What actually happened? Please include the actual, raw output from ESLint.**

It appears that when ESLint validates the **.eslint.json** file, it rejects the `$schema` field.  It prints this error:
```
Error: ESLint configuration in .eslintrc.json is invalid:
        - Unexpected top-level property "$schema".

    at validateConfigSchema (D:\GitRepos\eslint-test\node_modules\eslint\lib\shared\config-validator.js:286:15)
    at ConfigArrayFactory._normalizeConfigData (D:\GitRepos\eslint-test\node_modules\eslint\lib\cli-engine\config-array-factory.js:469:9)
    at ConfigArrayFactory._loadConfigDataInDirectory (D:\GitRepos\eslint-test\node_modules\eslint\lib\cli-engine\config-array-factory.js:445:33)
    at ConfigArrayFactory.loadInDirectory (D:\GitRepos\eslint-test\node_modules\eslint\lib\cli-engine\config-array-factory.js:401:18)
    at CascadingConfigArrayFactory._loadConfigInAncestors (D:\GitRepos\eslint-test\node_modules\eslint\lib\cli-engine\cascading-config-array-factory.js:304:46)
    at CascadingConfigArrayFactory.getConfigArrayForFile (D:\GitRepos\eslint-test\node_modules\eslint\lib\cli-engine\cascading-config-array-factory.js:249:18)
    at FileEnumerator._iterateFilesRecursive (D:\GitRepos\eslint-test\node_modules\eslint\lib\cli-engine\file-enumerator.js:408:49)
    at _iterateFilesRecursive.next (<anonymous>)
    at FileEnumerator.iterateFiles (D:\GitRepos\eslint-test\node_modules\eslint\lib\cli-engine\file-enumerator.js:254:49)
    at iterateFiles.next (<anonymous>)
```

**Are you willing to submit a pull request to fix this bug?**

Yes.  The simplest fix is for ESLint to delete the `$schema` field after it loads the file.

If people agree that we should fix this, I'm happy to create the PR.

**Comment by kaicataldo** [comment]
I don't know if this should be categorized as a bug - this seems like an editor-specific issue - but it does seem like something we should allow for if it's a common use case. I don't think we should delete the key, particularly for config files that aren't JSON files. What if the `$schema` property was allowed just for `.eslintrc.json` files (and would still cause a warning for a config file with any other extension)?

**Comment by octogonz** [comment]
> What if the `$schema` property was allowed just for `.eslintrc.json` files (and would still cause a warning for a config file with any other extension)?

That was what I meant.

However if I remember right, YAML also supports `$schema`.  Here's [an example](https://github.com/spacetelescope/yaml-schema-standard/blob/c575b2654360abd7353c599ed31944765e94d8cf/schemas/stsci.edu/yaml-schema/draft-01.yaml#L3).

BTW are people really accustomed to editing config files without any error-underlining or doc tooltips? This seems like a pretty basic feature.

**Comment by kaicataldo** [comment]
I personally have never encountered this before or seen it in any other JS open source project's config. Most editors will alert you to invalid JSON, but it sounds like this does more than that.

**Comment by octogonz** [comment]
It's a pretty cool feature!  You get tab completion when typing:

![json-intellisense](https://user-images.githubusercontent.com/4673363/61585714-aad20980-ab17-11e9-8e83-3521cadeba92.png)

...and error-underlining if you type it wrong:
![json-error](https://user-images.githubusercontent.com/4673363/61585715-b1608100-ab17-11e9-87f3-64eacc2ad0a4.png)

...and documentation tooltips when you mouse over a setting:
![json-tooltip](https://user-images.githubusercontent.com/4673363/61585716-ba515280-ab17-11e9-8073-bea45e55e69f.png)

**Comment by mysticatea** [comment]
I don't intend to oppose this enhancement, but as just a noticed, http://json.schemastore.org/eslintrc looks pretty old.

**Comment by leohxj** [comment]
Does eslint will support this feature?

**Comment by leohxj** [comment]
@octogonz What is the 'executable code', can they let eslint config work with `$schema`?
